test/performance/odp_l2fwd_perf.c

```diff
@@ -0,0 +1,1271 @@
+	    sigaction(SIGTERM, &action, NULL) == -1 || sigaction(SIGHUP, &action, NULL) == -1) {
+		ODPH_ERR("Error installing signal handler\n");
+		return PRS_NOK;
+	}
+
+	return parse_options(argc, argv, config);
+}
+
+static odp_bool_t setup_worker_config(prog_config_t *config)
+{
+	const uint32_t num = config->num_ifs;
+	pktio_t *pktio_src;
+	const pktio_t *pktio_dst;
+
+	if (config->mode == SCHED) {
+		if (odp_schedule_config(NULL) < 0) {
+			ODPH_ERR("Error initializing scheduler\n");
+			return false;
+		}
+	}
+
+	for (uint32_t i = 0U; i < num; ++i) {
+		pktio_src = &config->pktios[i];
+
+		for (uint32_t j = 0U; j < num; ++j) {
+			pktio_dst = &config->pktios[j];
+
+			if (pktio_src->dst_name != NULL &&
+			    strcmp(pktio_src->dst_name, pktio_dst->name) == 0) {
+				pktio_src->dst = pktio_dst;
+				break;
+			}
+		}
+	}
+
+	for (uint32_t i = 0U; i < config->num_workers; ++i)
+		config->worker_config[i].prog_config = config;
+
+	return true;
+}
+
+static odp_bool_t setup_pktios(prog_config_t *config)
+{
+	odp_pool_param_t pool_param;
+	pktio_t *pktio;
+	odp_pktio_param_t pktio_param;
+	odp_pktio_config_t pktio_config;
+	odp_pktio_capability_t capa;
+	uint32_t num_input_qs, num_output_qs;
+	odp_pktin_queue_param_t pktin_param;
+	odp_pktout_queue_param_t pktout_param;
+	int ret;
+
+	odp_pool_param_init(&pool_param);
+	pool_param.pkt.seg_len = config->pkt_len;
+	pool_param.pkt.len = config->pkt_len;
+	pool_param.pkt.num = config->num_pkts;
+	pool_param.type = ODP_POOL_PACKET;
+	config->pool = odp_pool_create(PROG_NAME, &pool_param);
+
+	if (config->pool == ODP_POOL_INVALID) {
+		ODPH_ERR("Error creating packet I/O pool\n");
+		return false;
+	}
+
+	for (uint32_t i = 0U; i < config->num_ifs; ++i) {
+		pktio = &config->pktios[i];
+		odp_pktio_param_init(&pktio_param);
+		pktio_param.in_mode = config->mode == DIRECT ?
+					ODP_PKTIN_MODE_DIRECT : ODP_PKTIN_MODE_SCHED;
+		pktio_param.out_mode = ODP_PKTOUT_MODE_DIRECT;
+		pktio->handle = odp_pktio_open(pktio->name, config->pool, &pktio_param);
+
+		if (pktio->handle == ODP_PKTIO_INVALID) {
+			ODPH_ERR("Error opening packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+
+		odp_pktio_config_init(&pktio_config);
+		pktio_config.parser.layer = ODP_PROTO_LAYER_NONE;
+
+		if (odp_pktio_config(pktio->handle, &pktio_config) < 0) {
+			ODPH_ERR("Error configuring packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+
+		if (odp_pktio_capability(pktio->handle, &capa) < 0) {
+			ODPH_ERR("Error querying packet I/O capabilities (%s)\n", pktio->name);
+			return false;
+		}
+
+		num_input_qs = ODPH_MIN((uint32_t)config->num_workers, capa.max_input_queues);
+		num_input_qs = ODPH_MIN(num_input_qs, MAX_QS);
+		num_output_qs = ODPH_MIN((uint32_t)config->num_workers, capa.max_output_queues);
+		num_output_qs = ODPH_MIN(num_output_qs, MAX_QS);
+		odp_pktin_queue_param_init(&pktin_param);
+
+		if (num_input_qs > 1) {
+			pktin_param.hash_enable = true;
+			pktin_param.hash_proto.proto.ipv4 = 1U;
```

Comment:
This hash protocol is not supported by the Intel XL710 NIC I'm using. So perhaps it's safer to continue using `ipv4_udp` as the  original l2fwd application does.